src/AdoNetCore.AseClient/AseDecimal.cs

```diff
@@ -0,0 +1,564 @@
+            var bigInt = new BigInteger(value);
+            Backing = isPositive
+                ? new BigDecimal(bigInt, -scale)
+                : new BigDecimal(-bigInt, -scale);
+        }
+
+        internal AseDecimal(BigDecimal backing)
+        {
+            Backing = backing;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the AseDecimal structure.
+        /// </summary>
+        /// <param name="value">A decimal structure to initialize the value of the new AseDecimal.</param>
+        public AseDecimal(decimal value)
+        {
+            Backing = value;
+        }
+
+        /// <summary>
+        /// Checks the sign of the AseDecimal value.
+        /// </summary>
+        /// <param name="value">The AseDecimal structure whose design is being checked.</param>
+        /// <returns>Returns an integer 0 for null, 1 for positive, and -1 for a negative values respectively.</returns>
+        public static int Sign(AseDecimal value)
+        {
+            return value.IsNull
+                ? 0
+                : value.IsNegative
+                    ? -1
+                    : 1;
+        }
+
+        /// <summary>
+        /// Returns a value indicating whether this instance of AseDecimal and the value indicated by the object value are equal.
+        /// </summary>
+        /// <param name="value">The object being tested for equality.</param>
+        public override bool Equals(object value)
+        {
+            if (ReferenceEquals(null, value))
+            {
+                return false;
+            }
+
+            return value is AseDecimal aseDecimal &&
+                   Backing == aseDecimal.Backing;
+        }
+
+        internal byte[] BinData
+        {
+            get
+            {
+                var result = new byte[BytesRequired];
+                var mantissaBytes = (Backing.Mantissa * Backing.Mantissa.Sign).ToByteArray();
+
+                Logger.Instance?.WriteLine($"Exponent: {Backing.Exponent}, Mantissa: {Backing.Mantissa}");
+                Logger.Instance?.WriteLine($"Expecting {BytesRequired} bytes; Mantissa is {mantissaBytes.Length} bytes");
+                Logger.Instance?.WriteLine($"Mantissa bytes:");
+                Logger.Instance?.Write(HexDump.Dump(mantissaBytes));
+
+                Array.Copy(mantissaBytes, 0, result, 0, Math.Min(result.Length, mantissaBytes.Length));
+                Logger.Instance?.WriteLine($"Result bytes:");
+                Logger.Instance?.Write(HexDump.Dump(result));
+                return result;
+            }
+        }
+
+        private static readonly double Log256 = Math.Log10(256);
+
+        /// <summary>
+        /// get the number of bytes required to represent the value (caller should add 1 byte to represent sign byte)
+        /// </summary>
+        public int BytesRequired
+        {
+            get
+            {
+                if (IsNull)
+                    throw new NullReferenceException();
```

Comment:
@senseibaka  - seeing an explicit throw of NullReferenceException made me vomit a little in the back of my mouth.

Reply:
@sheikhjabootie I've always wanted to do that 😄 🤢